test/validation/api/dma/dma.c

```diff
@@ -1603,6 +1796,211 @@ static void test_dma_pkt_to_pkt_multiple_offsets(void)
+
+	CU_ASSERT_FATAL(dma != ODP_DMA_INVALID);
+
+	return dma;
+}
+
+static odp_pool_t create_pool(uint32_t num)
+{
+	odp_pool_param_t pool_param = global.pool_param;
+	odp_pool_t pool;
+
+	pool_param.pkt.num = num;
+	pool = odp_pool_create("pkt_to_pkt_opts", &pool_param);
+
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+
+	return pool;
+}
+
+static uint32_t allocate_packets(odp_pool_t pool, odp_packet_t pkts[], uint32_t len, uint32_t num)
+{
+	odp_packet_t pkt;
+	uint32_t seg_len = 0;
+
+	for (uint32_t i = 0; i < num; i++) {
+		pkt = odp_packet_alloc(pool, len);
+
+		CU_ASSERT_FATAL(pkt != ODP_PACKET_INVALID);
+
+		seg_len = odp_packet_seg_len(pkt);
+		init_source(odp_packet_data(pkt), seg_len);
+		pkts[i] = pkt;
+	}
+
+	return seg_len;
+}
+
+static void setup_segs(odp_dma_seg_t segs[], odp_packet_t pkt, uint32_t len, uint32_t num,
+		       odp_bool_t use_idx)
+{
+	odp_dma_seg_t *seg;
+	const uint32_t seg_len = len / num;
+	uint32_t offset = 0;
+
+	for (uint32_t i = 0; i < num; i++) {
+		seg = &segs[i];
+
+		if (use_idx) {
+			seg->pkt_index = 0;
+			seg->pkt_len = len;
+		} else {
+			seg->packet = pkt;
+		}
+
+		seg->len = seg_len;
+		seg->offset = offset;
+		offset += seg_len;
+	}
+}
+
+static void test_dma_pkt_to_pkt_opts(odp_dma_compl_mode_t mask, const opts_t *opts)
+{
+	odp_dma_t dma;
+	odp_pool_t pool = global.pkt_pool;
+	/* One source and one destination packet. */
+	const uint32_t num_pkts = 2, src_idx = 0, dst_idx = 1,
+	num_src_seg = opts->unique_src_segs ? 1 : 2, num_dst_seg = opts->unique_dst_segs ? 1 : 2;
+	odp_packet_t pkts[num_pkts], pkt;
+	uint32_t seg_len;
+	odp_dma_seg_t src_seg[num_pkts];
+	odp_dma_seg_t dst_seg[num_pkts];
+	odp_dma_transfer_param_t trs_param;
+	int ret;
+
+	memset(src_seg, 0, sizeof(src_seg));
+	memset(dst_seg, 0, sizeof(dst_seg));
+	dma = create_dma(mask);
+
+	if (opts->seg_free)
+		pool = create_pool(num_pkts);
+
+	seg_len = allocate_packets(pool, pkts, global.pkt_len, num_pkts);
+	setup_segs(src_seg, pkts[src_idx], seg_len, num_src_seg, false);
+	setup_segs(dst_seg, pkts[dst_idx], seg_len, num_dst_seg, opts->seg_alloc);
+
+	odp_dma_transfer_param_init(&trs_param);
+	trs_param.src_format = ODP_DMA_FORMAT_PACKET;
+	trs_param.dst_format = ODP_DMA_FORMAT_PACKET;
+	trs_param.num_src = num_src_seg;
+	trs_param.num_dst = num_dst_seg;
+	trs_param.src_seg = src_seg;
+	trs_param.dst_seg = dst_seg;
+	trs_param.opts.seg_free = opts->seg_free;
+	trs_param.opts.unique_src_segs = opts->unique_src_segs;
+	trs_param.opts.single_pool = opts->single_pool;
+	trs_param.opts.seg_alloc = opts->seg_alloc;
+	trs_param.opts.unique_dst_segs = opts->unique_dst_segs;
+
+	if (opts->seg_alloc)
+		trs_param.dst_seg_pool = global.pkt_pool;
+
+	if (mask == ODP_DMA_COMPL_SYNC)
+		ret = do_transfer(dma, &trs_param, 0, 1);
+	else
+		ret = do_transfer_async(dma, &trs_param, mask, 0);
+
+	if (ret == 1) {
+		CU_ASSERT(check_equal(odp_packet_data(pkts[src_idx]),
+				      odp_packet_data(pkts[dst_idx]), seg_len) == 0);
```

Comment:
Is source packet already freed here if `seg_free` is enabled?

Reply:
Good catch, potentially already freed here. This whole check is a bit flawed anyways, have to rethink it.

Reply:
Reworked in v2.